src/backend/distributed/commands/function.c

```diff
@@ -0,0 +1,129 @@
+#include "distributed/metadata/distobject.h"
+#include "distributed/multi_executor.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/worker_transaction.h"
+#include "server/access/xact.h"
+#include "utils/builtins.h"
+#include "utils/fmgrprotos.h"
+
+/* forward declaration for helper functions*/
+static const char * GetFunctionDDLCommand(Oid funcOid);
+static void EnsureSequentialModeForFunctionDDL(void);
+
+PG_FUNCTION_INFO_V1(create_distributed_function);
+
+/*
+ * create_distributed_function gets a function or procedure name with their list of
+ * argument types in parantheses, then it creates a new distributed function.
+ */
+Datum
+create_distributed_function(PG_FUNCTION_ARGS)
+{
+	RegProcedure funcOid = PG_GETARG_OID(0);
+	const char *ddlCommand = NULL;
+	ObjectAddress functionAddress = { 0 };
+
+	/* if called on NULL input, error out */
+	if (funcOid == InvalidOid)
+	{
+		ereport(ERROR, (errmsg("create_distributed_function() requires a single "
+							   "parameter that is a valid function or procedure name "
+							   "followed by a list of parameters in parantheses"),
+						errhint("skip the parameters with OUT argtype as they are not "
+								"part of the signature in PostgreSQL")));
+	}
+
+	ObjectAddressSet(functionAddress, ProcedureRelationId, funcOid);
+
+	/*
+	 * when we allow propagation within a transaction block we should make sure to only
+	 * allow this in sequential mode
+	 */
+	EnsureSequentialModeForFunctionDDL();
+
+	EnsureDependenciesExistsOnAllNodes(&functionAddress);
+
+	ddlCommand = GetFunctionDDLCommand(funcOid);
+	SendCommandToWorkersAsUser(ALL_WORKERS, ddlCommand, NULL);
+
+	MarkObjectDistributed(&functionAddress);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetFunctionDDLCommand returns the complete "CREATE OR REPLACE FUNCTION ..." statement for
+ * the specified function.
+ */
+static const char *
+GetFunctionDDLCommand(RegProcedure funcOid)
+{
+	Datum sqlTextDatum = DirectFunctionCall1(pg_get_functiondef,
+											 ObjectIdGetDatum(funcOid));
+	const char *sql = TextDatumGetCString(sqlTextDatum);
+	return sql;
+}
+
+
+/*
+ * EnsureSequentialModeForFunctionDDL makes sure that the current transaction is already in
+ * sequential mode, or can still safely be put in sequential mode, it errors if that is
+ * not possible. The error contains information for the user to retry the transaction with
+ * sequential mode set from the beginnig.
+ *
+ * As functions are node scoped objects there exists only 1 instance of the function used by
+ * potentially multiple shards. To make sure all shards in the transaction can interact
+ * with the function the function needs to be visible on all connections used by the transaction,
+ * meaning we can only use 1 connection per node.
+ */
+static void
+EnsureSequentialModeForFunctionDDL(void)
+{
+	if (!IsTransactionBlock())
+	{
+		/* we do not need to switch to sequential mode if we are not in a transaction */
+		return;
```

Comment:
Let's remove this `if` block. It creates subtle bugs because multiple statements can be in the same transaction without being in a transaction block.

Reply:
I removed it. I used the logic we had around types.

I do not yet see the reason why we need a different behavior here. I will look into this further later.

The original if block:
https://github.com/citusdata/citus/blob/482f3b147452fff275663f649519b9e871b2036a/src/backend/distributed/commands/type.c#L1354-L1358

Reply:
Yep, it's wrong there too.